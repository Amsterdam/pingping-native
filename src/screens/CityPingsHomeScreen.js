import React from 'react';
import {RefreshControl, StatusBar, StyleSheet, View} from 'react-native';
import PropTypes from 'prop-types';
import {Container, Content, Header, Tab, Tabs} from 'native-base';
import {View as AnimatableView} from 'react-native-animatable';
import {useLazyQuery} from '@apollo/client';
import ContentLayout from '../components/layout/ContentLayout';
import Title from '../components/typography/Title';
import RewardCard from '../components/reward/RewardCard';
import {appColors} from '../config/colors';
import CitypingsChip from '../components/shared/CitypingsChip';
import ClaimedRewardsList from '../components/reward/ClaimedRewardsList';
import GET_AVAILABLE_REWARDS from '../apollo/Query/getAvailableRewards';
import GET_STATUS_QUERY from '../apollo/Query/getStatusQuery';
import ErrorComponent from '../components/shared/ErrorComponent';
import ClaimedRewardsModal from '../components/modals/ClaimedRewardModal';
import CardSkeleton from '../components/skeleton/CardSkeleton';

const CityPingsHomeScreen = ({navigation}) => {
  React.useEffect(() => {
    const unsubscribe = navigation.addListener('focus', () => {
      getAvailableRewards();
      getStatus();
    });
    return unsubscribe;
  }, [navigation, getAvailableRewards, getStatus]);

  const [getAvailableRewards, availableRewards] = useLazyQuery(
    GET_AVAILABLE_REWARDS,
  );
  const [getStatus, me] = useLazyQuery(GET_STATUS_QUERY, {
    fetchPolicy: 'cache-and-network',
  });
  const [refreshing, setRefreshing] = React.useState(false);

  const onRefresh = () => {
    setRefreshing(true);
    me.refetch();
    setRefreshing(false);
  };

  const retry = async () => {
    await availableRewards.refetch();
    await me.refetch();
  };

  if (availableRewards.error || me.error) {
    return (
      <ErrorComponent
        functionToRetry={retry}
        somethingWentWrong
        deafultLabelOverRide="Probeer Opnieuw"
        onPress={availableRewards.refetch}
      />
    );
  }

  const balance = me?.data?.getStatus?.user.balance; // maybe move this part to either localstate or the tabnavigator
  const claimedRewards = me?.data?.getStatus?.user.rewards;

  return (
    <Container style={styles.container}>
      <Header style={styles.header} transparent noShadow hasTabs>
        <StatusBar
          backgroundColor={appColors.primary}
          barStyle="light-content"
        />
        <View style={styles.headerContainer}>
          <Title style={styles.title}>Rewards</Title>
          <CitypingsChip value={balance} />
        </View>
      </Header>
      <Tabs
        tabBarUnderlineStyle={styles.tabBarUnderlineStyle}
        tabContainerStyle={styles.shadowRemover}>
        <Tab heading="Rewards" {...TAB_STYLE}>
          <Content
            style={{backgroundColor: appColors.almostNotBlue}}
            refreshControl={
              <RefreshControl
                refreshing={refreshing}
                onRefresh={onRefresh}
                tintColor={appColors.primary}
              />
            }>
            <ContentLayout>
              {availableRewards.loading && (
                <AnimatableView animation="fadeIn">
                  <CardSkeleton withTitle={false} />
                </AnimatableView>
              )}
              {availableRewards.data && (
                <AnimatableView animation="fadeIn">
                  {availableRewards.data.getAvailableRewards.map((reward) => (
                    <RewardCard
                      navigation={navigation}
                      reward={reward}
                      key={reward.rewardId}
                      balance={balance}
                    />
                  ))}
                </AnimatableView>
              )}
            </ContentLayout>
          </Content>
        </Tab>
        <Tab heading="Geclaimed" {...TAB_STYLE}>
          <ClaimedRewardsList
            claimedRewards={claimedRewards}
            navigation={navigation}
            balance={balance}
            getStatus={getStatus}
            onRefresh={onRefresh}
            refreshing={refreshing}
          />
        </Tab>
      </Tabs>
      <ClaimedRewardsModal navigation={navigation} />
    </Container>
  );
};

const styles = StyleSheet.create({
  header: {
    flexDirection: 'column',
    margin: 15,
    backgroundColor: appColors.primary,
    height: 100,
  },
  container: {backgroundColor: appColors.almostNotBlue},
  title: {
    color: '#fff',
    fontSize: 24,
    textAlign: 'left',
  },
  headerContainer: {
    flexDirection: 'row',
    justifyContent: 'space-between',
  },
  tabBarUnderlineStyle: {
    backgroundColor: '#fff',
    borderTopRightRadius: 10,
    borderTopLeftRadius: 10,
  },
  tabStyle: {
    backgroundColor: appColors.primary,
  },
  textStyle: {
    color: '#fff',
    fontFamily: 'Heavitas',
    fontSize: 12,
  },
  activeTextStyle: {
    color: '#fff',
    fontFamily: 'Heavitas',
    fontSize: 14,
  },
  shadowRemover: {
    elevation: 0,
  },
});

const TAB_STYLE = {
  tabStyle: styles.tabStyle,
  activeTabStyle: styles.tabStyle,
  textStyle: styles.textStyle,
  activeTextStyle: styles.activeTextStyle,
  backgroundColor: appColors.almostNotBlue,
};

CityPingsHomeScreen.propTypes = {
  navigation: PropTypes.object.isRequired,
};

export default CityPingsHomeScreen;
