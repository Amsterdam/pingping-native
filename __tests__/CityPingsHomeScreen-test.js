import React from 'react';
import CityPingsHomeScreen from '../src/screens/CityPingsHomeScreen';

import renderer from 'react-test-renderer';

test('renders correctly', () => {
  const tree = renderer.create(<CityPingsHomeScreen />).toJSON();
  expect(tree).toMatchSnapshot();
});
