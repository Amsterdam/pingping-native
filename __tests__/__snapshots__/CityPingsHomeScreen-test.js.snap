// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`renders correctly 1`] = `
<View
  style={
    Array [
      Object {
        "backgroundColor": "#fff",
        "flex": 1,
        "height": 1334,
      },
      Object {
        "backgroundColor": "#f7fbff",
      },
    ]
  }
>
  <View>
    <RCTSafeAreaView
      emulateUnlessSupported={true}
      style={
        Object {
          "backgroundColor": "#fb9f00",
        }
      }
    >
      <View
        hasTabs={true}
        noShadow={true}
        style={
          Array [
            Object {
              "backgroundColor": "transparent",
              "borderBottomColor": "transparent",
              "borderBottomWidth": null,
              "elevation": 0,
              "flexDirection": "row",
              "height": 64,
              "justifyContent": "center",
              "left": 0,
              "paddingLeft": 6,
              "paddingRight": 10,
              "paddingTop": 18,
              "right": 0,
              "shadowColor": null,
              "shadowOffset": null,
              "shadowOpacity": null,
              "shadowRadius": null,
              "top": 0,
            },
            Object {
              "backgroundColor": "#fb9f00",
              "flexDirection": "column",
              "height": 100,
              "margin": 15,
            },
          ]
        }
        transparent={true}
      >
        <View
          style={
            Object {
              "flexDirection": "row",
              "justifyContent": "space-between",
            }
          }
        >
          <Text
            style={
              Array [
                Object {
                  "color": "#000",
                  "fontFamily": "System",
                  "fontSize": 16,
                },
                Object {
                  "color": "#fff",
                  "fontFamily": "Heavitas",
                  "fontSize": 24,
                  "textAlign": "left",
                },
              ]
            }
            uppercase={false}
          >
            Rewards
          </Text>
          <View
            style={
              Object {
                "alignItems": "center",
                "backgroundColor": "#0d2036",
                "borderRadius": 5,
                "flexDirection": "row",
                "justifyContent": "center",
                "padding": 10,
              }
            }
          >
            <RNSVGSvgView
              align="xMidYMid"
              bbHeight={16}
              bbWidth={16}
              fill="none"
              focusable={false}
              height={16}
              meetOrSlice={0}
              minX={0}
              minY={0}
              style={
                Array [
                  Object {
                    "backgroundColor": "transparent",
                    "borderWidth": 0,
                  },
                  Object {
                    "marginRight": 5,
                  },
                  Object {
                    "flex": 0,
                    "height": 16,
                    "width": 16,
                  },
                ]
              }
              vbHeight={16}
              vbWidth={16}
              width={16}
            >
              <RNSVGGroup
                fill={null}
                propList={
                  Array [
                    "fill",
                  ]
                }
              >
                <RNSVGPath
                  clipRule={0}
                  d="M7.955 15.91a7.955 7.955 0 110-15.91 7.955 7.955 0 010 15.91z"
                  fill={4294892880}
                  fillRule={0}
                  propList={
                    Array [
                      "fill",
                      "fillRule",
                    ]
                  }
                />
                <RNSVGPath
                  clipRule={0}
                  d="M13.876 7.829a6.047 6.047 0 11-12.094 0 6.047 6.047 0 0112.094 0z"
                  fill={4294945536}
                  fillRule={0}
                  propList={
                    Array [
                      "fill",
                      "fillRule",
                    ]
                  }
                />
                <RNSVGPath
                  clipRule={0}
                  d="M14.047 8.08a5.966 5.966 0 11-11.932.001 5.966 5.966 0 0111.932 0z"
                  fill={4294950409}
                  fillRule={0}
                  propList={
                    Array [
                      "fill",
                      "fillRule",
                    ]
                  }
                />
                <RNSVGPath
                  clipRule={0}
                  d="M6.454 10.626c.425 0 .796-.076 1.113-.227.317-.152.586-.366.805-.641l-.707-.497a1.382 1.382 0 01-.497.417 1.531 1.531 0 01-.707.157c-.257 0-.488-.044-.693-.133a1.525 1.525 0 01-.528-.374 1.674 1.674 0 01-.34-.578 2.21 2.21 0 01-.119-.742c0-.257.04-.495.123-.714.081-.22.198-.41.35-.574a1.64 1.64 0 01.542-.385 1.705 1.705 0 011.316-.018c.21.082.38.212.511.389l.672-.532a1.442 1.442 0 00-.357-.336 2.406 2.406 0 00-.448-.234 2.528 2.528 0 00-.497-.14 3.033 3.033 0 00-1.561.136c-.327.121-.61.294-.85.518-.24.224-.43.496-.567.816a2.69 2.69 0 00-.207 1.074c0 .387.065.74.196 1.06.13.32.313.596.546.827.233.23.512.41.837.539.324.128.68.192 1.067.192zm3.596-.126V8.449h.679c.261 0 .509-.026.742-.077.233-.051.438-.134.613-.248.175-.115.313-.265.416-.452.103-.187.154-.418.154-.693a1.49 1.49 0 00-.133-.658 1.172 1.172 0 00-.378-.448c-.163-.117-.36-.2-.588-.252a3.525 3.525 0 00-.77-.077H9.168V10.5h.882zm0-2.793h.616c.145 0 .282-.01.413-.031.13-.022.247-.059.35-.112a.64.64 0 00.245-.221.658.658 0 00.091-.364.626.626 0 00-.09-.354.63.63 0 00-.242-.213 1.035 1.035 0 00-.343-.101 3.132 3.132 0 00-.396-.025h-.644v1.421z"
                  fill={4294960771}
                  fillRule={0}
                  propList={
                    Array [
                      "fill",
                      "fillRule",
                    ]
                  }
                />
              </RNSVGGroup>
            </RNSVGSvgView>
            <Text
              style={
                Array [
                  Object {
                    "color": "#000",
                    "fontFamily": "System",
                    "fontSize": 16,
                  },
                  Object {
                    "color": "#fff",
                    "fontFamily": "Heavitas",
                    "fontSize": 12,
                    "paddingTop": 5,
                    "textAlign": "left",
                  },
                ]
              }
              uppercase={false}
            >
              0
               CityPings
            </Text>
          </View>
        </View>
      </View>
    </RCTSafeAreaView>
  </View>
  <View
    onLayout={[Function]}
    style={
      Array [
        Object {
          "flex": 1,
        },
        undefined,
      ]
    }
  >
    <View
      style={
        Array [
          Object {
            "borderBottomWidth": 0.5,
            "borderColor": "#a7a6ab",
            "elevation": 3,
            "flexDirection": "row",
            "height": 50,
            "justifyContent": "space-around",
            "shadowColor": undefined,
            "shadowOffset": undefined,
            "shadowOpacity": undefined,
            "shadowRadius": undefined,
          },
          Object {
            "backgroundColor": "#F8F8F8",
          },
          Object {
            "elevation": 0,
          },
        ]
      }
    >
      <View
        accessible={true}
        focusable={true}
        onClick={[Function]}
        onResponderGrant={[Function]}
        onResponderMove={[Function]}
        onResponderRelease={[Function]}
        onResponderTerminate={[Function]}
        onResponderTerminationRequest={[Function]}
        onStartShouldSetResponder={[Function]}
        style={
          Object {
            "flex": 1,
            "opacity": 1,
          }
        }
      >
        <View
          active={true}
          style={
            Array [
              Object {
                "alignItems": "center",
                "backgroundColor": "#F8F8F8",
                "flex": 1,
                "flexDirection": "row",
                "justifyContent": "center",
              },
              Object {
                "backgroundColor": "#fb9f00",
              },
            ]
          }
        >
          <Text
            style={
              Array [
                Object {
                  "color": "#007aff",
                  "fontFamily": "System",
                  "fontSize": 16,
                  "fontWeight": "600",
                  "marginLeft": 7,
                  "marginRight": 7,
                },
                Object {
                  "fontSize": 15,
                },
                Object {
                  "color": "#fff",
                  "fontFamily": "Heavitas",
                  "fontSize": 14,
                },
                Object {
                  "color": "#fff",
                },
              ]
            }
            uppercase={false}
          >
            Rewards
          </Text>
        </View>
      </View>
      <View
        accessible={true}
        focusable={true}
        onClick={[Function]}
        onResponderGrant={[Function]}
        onResponderMove={[Function]}
        onResponderRelease={[Function]}
        onResponderTerminate={[Function]}
        onResponderTerminationRequest={[Function]}
        onStartShouldSetResponder={[Function]}
        style={
          Object {
            "flex": 1,
            "opacity": 1,
          }
        }
      >
        <View
          active={false}
          style={
            Array [
              Object {
                "alignItems": "center",
                "backgroundColor": "#F8F8F8",
                "flex": 1,
                "flexDirection": "row",
                "justifyContent": "center",
              },
              Object {
                "backgroundColor": "#fb9f00",
              },
            ]
          }
        >
          <Text
            style={
              Array [
                Object {
                  "color": "#6b6b6b",
                  "fontFamily": "System",
                  "fontSize": 16,
                  "marginLeft": 7,
                  "marginRight": 7,
                },
                Object {
                  "fontSize": 15,
                },
                Object {
                  "color": "#fff",
                  "fontFamily": "Heavitas",
                  "fontSize": 12,
                },
                Object {
                  "color": "#fff",
                },
              ]
            }
            uppercase={false}
          >
            Mijn prestaties
          </Text>
        </View>
      </View>
      <View
        style={
          Object {
            "backgroundColor": "#fff",
            "borderTopLeftRadius": 10,
            "borderTopRightRadius": 10,
            "bottom": 0,
            "height": 4,
            "left": 0,
            "marginHorizontal": 55,
            "position": "absolute",
            "width": 100,
          }
        }
      />
    </View>
    <RCTScrollView
      alwaysBounceVertical={false}
      automaticallyAdjustContentInsets={false}
      contentOffset={
        Object {
          "x": 0,
        }
      }
      directionalLockEnabled={true}
      horizontal={true}
      keyboardDismissMode="on-drag"
      keyboardShouldPersistTaps="handled"
      onMomentumScrollBegin={[Function]}
      onMomentumScrollEnd={[Function]}
      onScroll={[Function]}
      pagingEnabled={true}
      scrollEnabled={true}
      scrollEventThrottle={16}
      scrollsToTop={false}
      showsHorizontalScrollIndicator={false}
    >
      <View>
        <View
          style={
            Object {
              "width": 750,
            }
          }
        >
          <View
            activeTabStyle={
              Object {
                "backgroundColor": "#fb9f00",
              }
            }
            activeTextStyle={
              Object {
                "color": "#fff",
                "fontFamily": "Heavitas",
                "fontSize": 14,
              }
            }
            backgroundColor="#f7fbff"
            heading="Rewards"
            style={
              Object {
                "backgroundColor": "#FFF",
                "flex": 1,
              }
            }
            tabStyle={
              Object {
                "backgroundColor": "#fb9f00",
              }
            }
            textStyle={
              Object {
                "color": "#fff",
                "fontFamily": "Heavitas",
                "fontSize": 12,
              }
            }
          >
            <RCTSafeAreaView
              emulateUnlessSupported={true}
              style={
                Object {
                  "backgroundColor": "#f7fbff",
                  "flex": 1,
                }
              }
            >
              <RCTScrollView
                automaticallyAdjustContentInsets={false}
                contentContainerStyle={
                  Array [
                    Object {
                      "padding": undefined,
                    },
                    undefined,
                  ]
                }
                contentInset={
                  Object {
                    "bottom": 0,
                  }
                }
                enableAutomaticScroll={true}
                enableOnAndroid={false}
                enableResetScrollToCoords={true}
                extraHeight={75}
                extraScrollHeight={0}
                getScrollResponder={[Function]}
                handleOnScroll={[Function]}
                keyboardDismissMode="interactive"
                keyboardOpeningTime={250}
                keyboardShouldPersistTaps="handled"
                keyboardSpace={0}
                onScroll={[Function]}
                resetKeyboardSpace={[Function]}
                resetScrollToCoords={
                  Object {
                    "x": 0,
                    "y": 0,
                  }
                }
                scrollEventThrottle={1}
                scrollForExtraHeightOnAndroid={[Function]}
                scrollIntoView={[Function]}
                scrollToEnd={[Function]}
                scrollToFocusedInput={[Function]}
                scrollToPosition={[Function]}
                showsVerticalScrollIndicator={true}
                style={
                  Array [
                    Object {
                      "backgroundColor": "transparent",
                      "flex": 1,
                    },
                    Object {
                      "backgroundColor": "#f7fbff",
                    },
                  ]
                }
                update={[Function]}
                viewIsInsideTabBar={false}
              >
                <View>
                  <View
                    style={
                      Object {
                        "flex": 1,
                        "padding": 30,
                        "position": "relative",
                        "width": 750,
                      }
                    }
                  />
                </View>
              </RCTScrollView>
            </RCTSafeAreaView>
          </View>
        </View>
        <View
          style={
            Object {
              "width": 750,
            }
          }
        >
          <View
            heading="Mijn prestaties"
          />
        </View>
      </View>
    </RCTScrollView>
  </View>
</View>
`;
